### ISSUE TITLE
os: (*Process).Wait sometimes hangs on netbsd

### ISSUE THREAD

**Issue by bcmills** [golang/go#50138]
`greplogs --dashboard -md -l -e 'panic: test timed out.*\n\n(?:goroutine .*:\n(?:.+\n\t.+\n)+\n)*goroutine \d+ \[syscall, \d+ minutes\]:\n(?:.+\n\t.+\n)*os\.\(\*Process\)\.Wait(?:.*\n)+FAIL\s+cmd/link'`

[2021-12-12T06:14:07-9c6e8f6/netbsd-386-9_0-n2](https://build.golang.org/log/1b68bf9e1d56b6b9aa91de95711d0a7978277395)
```
goroutine 29 [syscall, 2 minutes]:
syscall.Syscall6(0x21a8, 0x892ee8c, 0x0, 0x8a2c1e0, 0x0, 0x0, 0x0)
	/tmp/workdir/go/src/syscall/asm_unix_386.s:43 +0x5 fp=0x892ee38 sp=0x892ee34 pc=0x80b9605
syscall.wait4(0x21a8, 0x892ee8c, 0x0, 0x8a2c1e0)
	/tmp/workdir/go/src/syscall/zsyscall_netbsd_386.go:34 +0x5b fp=0x892ee70 sp=0x892ee38 pc=0x80b737b
syscall.Wait4(0x21a8, 0x892eeb0, 0x0, 0x8a2c1e0)
	/tmp/workdir/go/src/syscall/syscall_bsd.go:144 +0x3b fp=0x892ee94 sp=0x892ee70 pc=0x80b558b
os.(*Process).wait(0x8a04660)
	/tmp/workdir/go/src/os/exec_unix.go:43 +0x82 fp=0x892eec8 sp=0x892ee94 pc=0x80de982
os.(*Process).Wait(...)
	/tmp/workdir/go/src/os/exec.go:132
os/exec.(*Cmd).Wait(0x8a18fd0)
	/tmp/workdir/go/src/os/exec/exec.go:507 +0x4d fp=0x892ef0c sp=0x892eec8 pc=0x816b07d
os/exec.(*Cmd).Run(0x8a18fd0)
	/tmp/workdir/go/src/os/exec/exec.go:341 +0x43 fp=0x892ef1c sp=0x892ef0c pc=0x816a463
os/exec.(*Cmd).CombinedOutput(0x8a18fd0)
	/tmp/workdir/go/src/os/exec/exec.go:567 +0x89 fp=0x892ef30 sp=0x892ef1c pc=0x816b549
cmd/link.TestContentAddressableSymbols(0x89290e0)
	/tmp/workdir/go/src/cmd/link/link_test.go:879 +0x136 fp=0x892ef9c sp=0x892ef30 pc=0x83824b6
testing.tRunner(0x89290e0, 0x842c054)
	/tmp/workdir/go/src/testing/testing.go:1410 +0x10d fp=0x892efe4 sp=0x892ef9c pc=0x813d19d
testing.(*T).Run.func1()
	/tmp/workdir/go/src/testing/testing.go:1457 +0x28 fp=0x892eff0 sp=0x892efe4 pc=0x813df78
runtime.goexit()
	/tmp/workdir/go/src/runtime/asm_386.s:1311 +0x1 fp=0x892eff4 sp=0x892eff0 pc=0x80ab211
created by testing.(*T).Run
	/tmp/workdir/go/src/testing/testing.go:1457 +0x36e
```
[2021-10-29T18:34:24-903f313/netbsd-amd64-9_0](https://build.golang.org/log/e4e0ddf202e3fee6729c2a4be5ce9945ec24015f)
[2021-10-01T15:59:38-e5ad363/netbsd-arm-bsiegert](https://build.golang.org/log/b1b3360fc6590c5f956bfd3b00060e97b08b7178)
```
goroutine 28 [syscall, 27 minutes]:
syscall.Syscall6(0x1c1, 0xd1f, 0xa09db4, 0x0, 0x9b27e0, 0x0, 0x0)
	/var/gobuilder/buildlet/go/src/syscall/asm_netbsd_arm.s:39 +0x8 fp=0xa09d5c sp=0xa09d58 pc=0x8d3f8
syscall.wait4(0xd1f, 0xa09db4, 0x0, 0x9b27e0)
	/var/gobuilder/buildlet/go/src/syscall/zsyscall_netbsd_arm.go:35 +0x54 fp=0xa09d94 sp=0xa09d5c pc=0x8a694
syscall.Wait4(0xd1f, 0xa09dd8, 0x0, 0x9b27e0)
	/var/gobuilder/buildlet/go/src/syscall/syscall_bsd.go:145 +0x3c fp=0xa09db8 sp=0xa09d94 pc=0x88c58
os.(*Process).wait(0x983290)
	/var/gobuilder/buildlet/go/src/os/exec_unix.go:44 +0x100 fp=0xa09df0 sp=0xa09db8 pc=0xb4f1c
os.(*Process).Wait(...)
	/var/gobuilder/buildlet/go/src/os/exec.go:132
os/exec.(*Cmd).Wait(0x98cc60)
	/var/gobuilder/buildlet/go/src/os/exec/exec.go:507 +0x50 fp=0xa09e2c sp=0xa09df0 pc=0x1482d0
os/exec.(*Cmd).Run(0x98cc60)
	/var/gobuilder/buildlet/go/src/os/exec/exec.go:341 +0x48 fp=0xa09e3c sp=0xa09e2c pc=0x147810
os/exec.(*Cmd).CombinedOutput(0x98cc60)
	/var/gobuilder/buildlet/go/src/os/exec/exec.go:567 +0x98 fp=0xa09e50 sp=0xa09e3c pc=0x14882c
cmd/link.TestIssue33979.func2({0x983200, 0x21}, {0x9ea0a0, 0x9, 0x9})
	/var/gobuilder/buildlet/go/src/cmd/link/link_test.go:199 +0x90 fp=0xa09ea8 sp=0xa09e50 pc=0x368e14
cmd/link.TestIssue33979.func3({0x9ea0a0, 0x9, 0x9})
	/var/gobuilder/buildlet/go/src/cmd/link/link_test.go:206 +0x60 fp=0xa09ecc sp=0xa09ea8 pc=0x368d5c
cmd/link.TestIssue33979(0x8834a0)
	/var/gobuilder/buildlet/go/src/cmd/link/link_test.go:239 +0x3bc fp=0xa09f98 sp=0xa09ecc pc=0x368790
testing.tRunner(0x8834a0, 0x41871c)
	/var/gobuilder/buildlet/go/src/testing/testing.go:1389 +0x118 fp=0xa09fe0 sp=0xa09f98 pc=0x1195d4
testing.(*T).Run.func1()
	/var/gobuilder/buildlet/go/src/testing/testing.go:1436 +0x30 fp=0xa09fec sp=0xa09fe0 pc=0x11a448
runtime.goexit()
	/var/gobuilder/buildlet/go/src/runtime/asm_arm.s:824 +0x4 fp=0xa09fec sp=0xa09fec pc=0x7d028
created by testing.(*T).Run
	/var/gobuilder/buildlet/go/src/testing/testing.go:1436 +0x3a0
```
[2021-09-21T20:39:31-48cf96c/netbsd-arm-bsiegert](https://build.golang.org/log/97dd6f9a69b7b435011158f97d49993856a8c4b6)
[2021-09-14T14:27:57-181e8cd/netbsd-arm-bsiegert](https://build.golang.org/log/105073818ca301f41164d2e8122e3a1e312d916f)
[2021-04-29T15:47:16-12eaefe/freebsd-amd64-11_4](https://build.golang.org/log/60725cc8e6ed2ba02a10b88a377a13fc413fb80a)
[2021-04-28T13:49:52-4fe324d/netbsd-386-9_0](https://build.golang.org/log/84d454b467e421ea3a24e618e41c5a0eac4f1d05)
[2021-03-05T02:30:31-b62da08/netbsd-386-9_0](https://build.golang.org/log/a90168a43a8ff7c7132c9c356d53b2868ad5d79a)
[2021-02-19T00:40:05-95a44d2/netbsd-arm64-bsiegert](https://build.golang.org/log/ecc915eeabf45bb8b03b97532af0ac20f3d07af1)
[2019-09-04T21:52:18-aae0b5b/linux-ppc64le-power9osu](https://build.golang.org/log/3a5a781ecc670b125d283ebe14ddd5ed575a666f)

#44801 may be closely related.

Note that many of this failures are on architectures <strong>not</strong> believed to be affected by #49209.

@bsiegert, @coypoop: any ideas?

**Comment by bcmills** [comment]
Could also be related to #48789.

**Comment by bcmills** [comment]
The stuck calls appear to be running `go run`, `go tool link`, etc. Since these are Go binaries, probably a good starting point would be to send the stuck processes `SIGQUIT` to try to get goroutine dumps, and then to send them `SIGKILL` if they still don't respond.

That would at least help us to determine whether the hang is in the subprocess or the parent process.

(I think @aclements and @mknyszek were working on retrofitting that logic to various tests?)

**Comment by cherrymui** [comment]
The arm and arm64 ones may be due to slow machine.

Yeah, sending a SIGQUIT at timeout is probably a good idea.

**Comment by bcmills** [comment]
Looks like the same failure mode in `go/internal/gcimporter` too: https://build.golang.org/log/0e1b9a393109ba16005d18ff9faca47c50728a8f

**Comment by bcmills** [comment]
`greplogs --dashboard -md -l -e 'panic: test timed out.*\n\n(?:goroutine .*:\n(?:.+\n\t.+\n)+\n)*goroutine \d+ \[syscall, \d+ minutes\]:\n(?:.+\n\t.+\n)*os\.\(\*Process\)\.Wait(?:.*\n)+FAIL\s+cmd/link' --since=2022-01-08`

[2022-03-10T09:12:04-5a040c5/netbsd-amd64-9_0](https://build.golang.org/log/b37a5bb392bc486713292601ee50f548ee92b7b2)

**Comment by bcmills** [comment]
Broadening the regexp to search for `os.(*Process).Wait` generally, since this symptom does not appear to be specific to `cmd/link` as far as I can see.

`greplogs --dashboard -md -l -e '\Anetbsd-(.*\n)*panic: test timed out.*\n\n(?:goroutine .*:\n(?:.+\n\t.+\n)+\n)*goroutine \d+ \[syscall, \d+ minutes\]:\n(?:.+\n\t.+\n)*os\.\(\*Process\)\.Wait' --since=2022-01-01`

* [2022-04-05T17:54:15-de8a882-9839668/netbsd-386-9_0](https://build.golang.org/log/3f564e11e78acc39c5cb531a481ae92c03d00f0c)
   * 9 minutes, stuck in `git-codereview.TestSubmit` on a `git fetch` to a local server hosted in the test process, and the server appears to be idle
* [2022-03-31T06:40:26-ebe624d/netbsd-arm-bsiegert](https://build.golang.org/log/3fa57fa03238a84c26034f296e12c0b41f19dee3)
   * 29 minutes, stuck in `gcimporter_test.TestImportTypeparamTests` on `go tool compile`
* [2022-03-10T09:12:04-5a040c5/netbsd-amd64-9_0](https://build.golang.org/log/b37a5bb392bc486713292601ee50f548ee92b7b2)
   * 2 minutes, stuck in `link.TestIssue21703` on `go tool compile`
* [2022-02-14T17:30:31-6c99688-3d7f836/netbsd-386-9_0](https://build.golang.org/log/4e25bee0c27843bd83854f6c6fa8460f45403d6c)
   * 9 minutes, same `git fetch` in `git-codereview.TestSubmit`
* [2022-01-13T19:25:50-ce01afe/netbsd-arm64-bsiegert](https://build.golang.org/log/5b00db52ab4b384a1ce490677eb7fd0009c54b18)
   * 29 minutes, stuck in `pprof.TestMapping` on `go run` of a test program that sleeps for 200ms before exiting
* [2022-01-11T15:28:40-c7fa661/netbsd-arm-bsiegert](https://build.golang.org/log/efbcd57b051c89f22213c5dabe825310bd965bad)
   * 29 minutes, stuck in `cmd/compile/internal/test.TestReproducibleBuilds` on `go tool compile`
* [2022-01-06T21:38:57-c5540e5/netbsd-386-9_0-n2](https://build.golang.org/log/0e1b9a393109ba16005d18ff9faca47c50728a8f)
   * 2 minutes, same `go tool compile` in `gcimporter_test.TestImportTypeparamTests`

**Comment by bcmills** [comment]
`greplogs --dashboard -md -l -e '\Anetbsd-(.*\n)*panic: test timed out.*\n\n(?:goroutine .*:\n(?:.+\n\t.+\n)+\n)*goroutine \d+ \[syscall, \d+ minutes\]:\n(?:.+\n\t.+\n)*os\.\(\*Process\)\.Wait' --since=2022-04-06`

[2022-04-18T22:07:54-f49e802/netbsd-arm-bsiegert](https://build.golang.org/log/232b3949ab3fd984c49944fcaf921037f19c61e4)
   * 29 minutes, stuck in `cmd/link/internal/ld.TestMemProfileCheck` on `go run` of a test program that prints `runtime.MemProfileRate` and then exits.

**Comment by coypoop** [comment]
> The arm and arm64 ones may be due to slow machine.

Is this still a suspicion? I assume the netbsd/arm builder is extra slow.
I didn't try this, but netbsd/arm64's 32-bit compat might be sufficient for running arm32 tests.

**Comment by riastradh** [comment]
Do you have steps to reproduce?

I'm having a little trouble following the initial report, because it seems to cover several operating systems and architectures.  Does this happen every time on NetBSD, or on NetBSD/arm, or only sometimes, or what?

If it happens only sometimes, how long does it take successful test runs on the machines where it fails?

**Comment by bcmills** [comment]
> > The arm and arm64 ones may be due to slow machine.
>
> Is this still a suspicion? I assume the netbsd/arm builder is extra slow.

I can't speak for @cherrymui, but given the similar failures on the `-amd64` and `-386` builders I believe this is likely an architecture-independent synchronization bug in either the NetBSD kernel or the Go `os` implementation for NetBSD.

**Comment by bcmills** [comment]
> Do you have steps to reproduce?

Unfortunately no. The failures listed above were found organically in the Go build dashboard — the repro rate is high enough to be significant but not high enough to reproduce on demand.

> I'm having a little trouble following the initial report, because it seems to cover several operating systems and architectures.

The freebsd failure in `2021-04-29T15:47:16-12eaefe` was likely #46272, and hasn't been seen since that was addressed.
The linux-ppc64le failure was from 2019, and may have been an organic timeout. (It hasn't been a recurring pattern.)
The remainder AFAICT are on NetBSD, on varying architectures.

> Does this happen every time on NetBSD, or on NetBSD/arm, or only sometimes, or what?

Intermittently, on NetBSD across all of the architectures for which we have builders.

**Comment by bcmills** [comment]
(Note that we also tried to use `wait6` for this on NetBSD, but had to roll it back because of other deadlocks on that OS — see #48789.)

**Comment by bcmills** [comment]
`greplogs --dashboard -md -l -e '\Anetbsd-(.*\n)*panic: test timed out.*\n\n(?:goroutine .*:\n(?:.+\n\t.+\n)+\n)*goroutine \d+ \[syscall, \d+ minutes\]:\n(?:.+\n\t.+\n)*os\.\(\*Process\)\.Wait' --since=2022-04-19`

* [2022-04-22T04:42:23-1e59876/netbsd-386-9_0](https://build.golang.org/log/10630fc7a09bb1808f3552aa54a312157d95396a)
   * 8 minutes, stuck in `cmd/go_test.waitOrStop`, which attempts to explicitly terminate the process before the observed timout occurs (see #50187, #48789 — both of which appear to be specific to NetBSD).
* [2022-04-22T00:37:23-bf19163/netbsd-386-9_0](https://build.golang.org/log/769e72c2adb459dea7f9c14749deb795082f1fae)
   * 2 minutes, stuck in `runtime_test.TestLockOSThreadTemplateThreadRace` on a [program](https://cs.opensource.google/go/go/+/master:src/runtime/testdata/testprog/lockosthread.go;l=201-246;drc=4abec2a48070da6ca9b8cf53888ad993e0ee82ef) that spins on `time.Now()` on two goroutines for 10ms before locking those threads and exiting.

**Comment by bcmills** [comment]
`greplogs --dashboard -md -l -e '\Anetbsd-(.*\n)*panic: test timed out.*\n\n(?:goroutine .*:\n(?:.+\n\t.+\n)+\n)*goroutine \d+ \[syscall, \d+ minutes\]:\n(?:.+\n\t.+\n)*os\.\(\*Process\)\.Wait' --since=2022-04-23`
[2022-05-12T22:29:02-da0a6f4/netbsd-arm-bsiegert](https://build.golang.org/log/117d494942b35f2da01399a69ac0289474f8caa4)
[2022-05-12T20:19:10-27ace7a/netbsd-arm-bsiegert](https://build.golang.org/log/7145d7805b0b912319cb5817d17a181238afd33d)

Very curious that these seem to occur in pairs. 🤔

(attn @golang/netbsd)

**Comment by bcmills** [comment]
`greplogs -l -e '\Anetbsd-(.*\n)*panic: test timed out.*\n\n(?:goroutine .*:\n(?:.+\n\t.+\n)+\n)*goroutine \d+ \[syscall, \d+ minutes\]:\n(?:.+\n\t.+\n)*os\.\(\*Process\)\.Wait' --since=2022-05-14`
[2022-05-27T14:57:14-590b53f/netbsd-amd64-9_0](https://build.golang.org/log/9ed2fef8113fa1ebdaf90f57547bd4e7818b2bcc) (on release-branch.go1.17)
[2022-05-17T03:26:28-41b9d8c/netbsd-arm64-bsiegert](https://build.golang.org/log/e0852b1fcb2375d635ef8cd80d6a025e8a28c9c7) (on the main branch)

**Comment by bcmills** [comment]
`greplogs -l -e '\Anetbsd-(.*\n)*panic: test timed out.*\n\n(?:goroutine .*:\n(?:.+\n\t.+\n)+\n)*goroutine \d+ \[syscall, \d+ minutes\]:\n(?:.+\n\t.+\n)*os\.\(\*Process\)\.Wait' --since=2022-05-31 --details`
<details>

[2022-09-04T04:17:04-535fe2b/netbsd-386-9_0](https://build.golang.org/log/7c945382a68a2104bb5b4414aff533ced1b2c44f)
[2022-09-02T05:05:13-33c1ddd-8e35910/netbsd-amd64-9_0](https://build.golang.org/log/2397999715ec6855f7d33e260e57985d8c41d1d7)
[2022-09-02T02:09:04-a330ca5/netbsd-386-9_0](https://build.golang.org/log/be6cda283947b7bf591a1449b6f74032ee0a7f02)
[2022-09-02T00:06:00-40cfaff-0592ce5/netbsd-amd64-9_0](https://build.golang.org/log/233feb2de27b6a09639940ac94e3d90732ef41ff)
[2022-09-01T23:18:06-40cfaff-1280ae7/netbsd-amd64-9_0](https://build.golang.org/log/fbe9587e4eda9a8fb9eb4783eb804d202d6d43bd)
[2022-09-01T22:37:04-40cfaff-aa5ff29/netbsd-386-9_0](https://build.golang.org/log/000aa3c8d952be65ee7ed1014757069e88edb640)
[2022-09-01T18:08:05-f16be35-ef84141/netbsd-386-9_0](https://build.golang.org/log/e8248340783cdc7c7fb313d588f367f2892f7cad)
[2022-09-01T15:27:30-6c10975-86e9e0e/netbsd-386-9_0](https://build.golang.org/log/994db372977e46635bc3a3264a9012ff5e1f5f86)
[2022-09-01T03:24:42-49ab44d-91ef076/netbsd-amd64-9_0](https://build.golang.org/log/781a919765bffee76e3feca038a28f7f12aa26b5)
[2022-09-01T00:42:27-550e1f5-64b260d/netbsd-386-9_0](https://build.golang.org/log/eb9903a5db58c2139b1a6bd29c33ce79d021957c)
[2022-08-31T23:52:00-550e1f5-ca634fa/netbsd-amd64-9_0](https://build.golang.org/log/5c06c96640447e25876152f3ff810f4e6b10d62b)
[2022-08-31T22:22:43-550e1f5-e4b624e/netbsd-amd64-9_0](https://build.golang.org/log/050b8ca1efebd1ea3a55e9dcad68c30d30c591b3)
[2022-08-31T22:10:52-550e1f5-33a7e5a/netbsd-386-9_0](https://build.golang.org/log/ccb1be631338df0b55cfaa360b1183aa7f3f6094)
[2022-08-31T21:08:24-4ccc73c-ce77a46/netbsd-386-9_0](https://build.golang.org/log/a6b0fd946e119e120c4871beb98236acbd801845)
[2022-08-31T21:08:24-4ccc73c-889d326/netbsd-amd64-9_0](https://build.golang.org/log/cc5ade97bdf85213ea8c0245486b7de03d5d65a2)
[2022-08-31T16:26:07-41c3a9b-d2d5929/netbsd-386-9_0](https://build.golang.org/log/4ec17b41eb3c5a8d9c4733d15d1b415b3b4235df)
[2022-08-31T01:16:54-248c34b-ee0e40a/netbsd-amd64-9_0](https://build.golang.org/log/04547691fcaf680bf02a081c4b6c57be6b634361)
[2022-08-30T21:30:15-248c34b-bd56cb9/netbsd-amd64-9_0](https://build.golang.org/log/6cc62a717ad7cf778406c971c15ea51c3bb759e1)
[2022-08-30T00:49:19-248c34b-629ae1c/netbsd-amd64-9_0](https://build.golang.org/log/4c80dc13469ab295d94bc578796f898ed0ae9f82)
[2022-08-28T16:04:49-717a671-846c378/netbsd-386-9_0](https://build.golang.org/log/c4a66b7d3f4782f6e7b57dfbdfe3d2238b75e648)
[2022-08-26T19:15:02-717a671-897ad2f/netbsd-amd64-9_0](https://build.golang.org/log/e21a233dc60e92caa9672a12bffe6b66838171a3)
[2022-08-26T18:28:14-717a671-bf812b3/netbsd-386-9_0](https://build.golang.org/log/6ef77844158653a3a5367aabbe2e0c3ad4a44456)
[2022-08-26T17:48:20-7f23307-296c40d/netbsd-amd64-9_0](https://build.golang.org/log/4cf3360c8a6818759645164f6bce37db1a11bdc2)
[2022-08-26T17:15:08-bc1d0d8/netbsd-amd64-9_0](https://build.golang.org/log/6b21f7762ad9606b0dc670c5a28ff2ac38d51966)
[2022-08-26T15:36:36-7c5e035-951d2c6/netbsd-386-9_0](https://build.golang.org/log/a4183f76efa7cee46b3f8f13fdaa862608cdb2c1)
[2022-08-25T19:17:14-d35bb19-f64f12f/netbsd-386-9_0](https://build.golang.org/log/1916bbd565f13badf52a10c6cbee34099ec8a384)
[2022-08-25T17:31:33-db6a62c-83b5fe6/netbsd-386-9_0](https://build.golang.org/log/26034ffccf387ce6d5da51317b7c0d5465405fca)
[2022-08-25T14:56:18-db6a62c-e4be2ac/netbsd-amd64-9_0](https://build.golang.org/log/080c80f427d6a3e9206f9af765253c5ff2518797)
[2022-08-25T04:00:07-db6a62c-8c8429f/netbsd-386-9_0](https://build.golang.org/log/0632c880b16579729dbe2491c489fc0e1b0d3c29)
[2022-08-25T04:00:07-db6a62c-8c8429f/netbsd-amd64-9_0](https://build.golang.org/log/a666ea4c99cdc4cf68f6f734c2b90571c3d83d97)
[2022-08-24T21:20:11-db6a62c-e4bed41/netbsd-amd64-9_0](https://build.golang.org/log/c8421302a1803c562b8e6e0110404d8c94f370a9)
[2022-08-24T21:09:24-e4bed41/netbsd-386-9_0](https://build.golang.org/log/9ce837e9714d3902a034d838b6fdae84d117e84e)
[2022-08-24T17:56:56-587a153-cfae70c/netbsd-386-9_0](https://build.golang.org/log/baa65ed728fd00f8c4e5480d0bda9188ffd101e7)
[2022-08-24T15:37:38-587a153-d5aa088/netbsd-amd64-9_0](https://build.golang.org/log/7e2dcd2a4f1f3077665da8083e60d87f0b1e4149)
[2022-08-24T14:31:08-c837a30-f983a93/netbsd-amd64-9_0](https://build.golang.org/log/b3d226bb3a00f766e2035ccf58638282c7a43c7c)
[2022-08-24T12:12:12-587a153-b5a9459/netbsd-amd64-9_0](https://build.golang.org/log/caecb7084bb781ba347f65fb6f57aac5e21fbfc8)
[2022-08-23T20:32:50-fc0d423/netbsd-386-9_0](https://build.golang.org/log/c76d6c4b08c74c0d766b970016a50a3e0bb3ff9e)
[2022-08-23T20:32:50-587a153-fc0d423/netbsd-386-9_0](https://build.golang.org/log/dc8a5042203b88507155c4d52b6b5ac6e07dbb47)
[2022-08-22T18:51:32-a726c9f/netbsd-386-9_0](https://build.golang.org/log/e63a492484001691a093627e6ebfd7bfb0788959)
[2022-08-22T16:48:36-a10da77/netbsd-amd64-9_0](https://build.golang.org/log/704d7d571a744f5055612ccd2050aa8e71ca3245)
[2022-08-19T16:55:03-e55fb40-7dad1d2/netbsd-386-9_0](https://build.golang.org/log/59471b61d9a3de3c219f037440229a8b2996dbc6)
[2022-08-19T16:17:50-e55fb40-5729419/netbsd-386-9_0](https://build.golang.org/log/ae365a434f2ad1aba86343dcbcf1c37b7e88776d)
[2022-08-19T15:53:47-e55fb40-f324355/netbsd-amd64-9_0](https://build.golang.org/log/404c696f3c2531de885e9feec09e0b88666ea99d)
[2022-08-19T03:09:05-e55fb40-a409356/netbsd-386-9_0](https://build.golang.org/log/5d933c6dbd4cbee571eb6891ba674bbfb2eecda7)
[2022-08-15T21:54:27-e3c2e4c/netbsd-386-9_0](https://build.golang.org/log/f552f18552312be7579b51212aac5914f229a187)
[2022-08-15T20:41:00-938e162-de0f4d1/netbsd-amd64-9_0](https://build.golang.org/log/814442b3c379948160d82e9310e011b0034d0320)
[2022-08-15T20:02:31-938e162-7b45edb/netbsd-amd64-9_0](https://build.golang.org/log/8b6fe466d2eebfb8af568bae2fcf62320186d6a2)
[2022-08-15T19:17:20-8c83056-7b45edb/netbsd-amd64-9_0](https://build.golang.org/log/03a8688ab5b21f6eacb68e25ffb8c7a521e98523)
[2022-08-15T15:15:10-f1b1557/netbsd-386-9_0](https://build.golang.org/log/8eb7c9543a8a6698199ec5874d5b4f860f56f874)
[2022-08-14T00:06:23-35f806b-59865f1/netbsd-amd64-9_0](https://build.golang.org/log/cac6093c9a1e29b020d1a0076f0b67e262a134c5)
[2022-08-12T20:40:05-bebd890-2f6783c/netbsd-386-9_0](https://build.golang.org/log/3680daecab8adb2905484af9298a4239b3812f6a)
[2022-08-12T18:15:28-bebd890-b6f87b0/netbsd-amd64-9_0](https://build.golang.org/log/c2900e262adb237ea981093befa87ec35faa804b)
[2022-08-12T12:39:26-88d981e-f67c766/netbsd-amd64-9_0](https://build.golang.org/log/d22326763f8cbedb39bb3c17f3e66e57efa453df)
[2022-08-12T00:04:29-c4ec74a-a5cd894/netbsd-386-9_0](https://build.golang.org/log/eec2b475f968a73f4f17745f72013ca6b3dc2a26)
[2022-08-11T20:16:35-6c2e327/netbsd-amd64-9_0](https://build.golang.org/log/c083d8b3c009108db2b6460a2ed4d2cb4c4928ab)
[2022-08-11T20:13:14-db84f53/netbsd-amd64-9_0](https://build.golang.org/log/9a17f94947d382f5c0b010712209d0240a5300d4)
[2022-08-11T17:53:50-37a81b6-a526ec1/netbsd-amd64-9_0](https://build.golang.org/log/373391707f93fff54945c55648170d9dd390958c)
[2022-08-11T16:19:14-37a81b6-2340d37/netbsd-amd64-9_0](https://build.golang.org/log/1c1abf0fdf7307ffbe030628c0c37a2af50e044d)
[2022-08-10T23:26:58-2e6ffd6/netbsd-amd64-9_0](https://build.golang.org/log/aa4d308d7b05fe44961d0805578e062cf0e98d64)
[2022-08-10T22:22:48-b2156b5-6b80b62/netbsd-amd64-9_0](https://build.golang.org/log/0e0cf575d88b65e91a55b8f070bf9b7ef3136e6e)
[2022-08-10T17:41:25-0ad49fd-f19f6c7/netbsd-amd64-9_0](https://build.golang.org/log/f1c0a3ff9bc966ba2ea145a3fc5b207e7da04422)
[2022-08-10T15:08:24-3950865-c81dfdd/netbsd-amd64-9_0](https://build.golang.org/log/49e68c6f9a3e02fea5a3560b3599fec748fe37a5)
[2022-08-09T14:33:24-0981d9f/netbsd-386-9_0](https://build.golang.org/log/952f5cbdf0616cbbcabcb5adceea5738eb05ede2)
[2022-08-09T14:12:01-92d58ea-662a729/netbsd-amd64-9_0](https://build.golang.org/log/262bb87feffe594abd0bca78d7ff0762bce6ce6d)
[2022-08-09T11:28:56-92d58ea-0f8dffd/netbsd-amd64-9_0](https://build.golang.org/log/6f487445444131ce88dc81395c61491a75b717b6)
[2022-08-08T18:14:49-a34a97d/netbsd-amd64-9_0](https://build.golang.org/log/c6685cae3900fec8e6aee2e62c8c29df77dd744e)
[2022-08-08T17:01:54-9b60852-0a86cd6/netbsd-amd64-9_0](https://build.golang.org/log/6626f50b373eac45d25ec68fbec811c676b240e3)
[2022-08-08T16:05:18-487b350/netbsd-386-9_0](https://build.golang.org/log/0a46791f295fb3cb3340c47a65e86eb0f8c9d04a)
[2022-08-08T15:07:46-06d96ee-cd54ef1/netbsd-amd64-9_0](https://build.golang.org/log/4bf9c1e647958f6f7e91f6a4af54f5c0d47a5ff9)
[2022-08-08T14:11:19-6e9925c/netbsd-amd64-9_0](https://build.golang.org/log/3b5f8d5ab9b55edc86aa62eab24e7c55f99cd365)
[2022-08-08T14:11:09-3a9281f/netbsd-amd64-9_0](https://build.golang.org/log/11287a66abd28461e3a98f27e33591770b12ca37)
[2022-08-08T06:16:59-06d96ee-0f6ee42/netbsd-386-9_0](https://build.golang.org/log/e338e8cc5f896780178f4653ca14cb3c5420eee3)
[2022-08-04T20:05:03-81c7dc4-39728f4/netbsd-386-9_0](https://build.golang.org/log/a0e0e2e24dbcc85bc1112cde350e2aab2f68507e)
[2022-08-04T20:05:03-81c7dc4-39728f4/netbsd-amd64-9_0](https://build.golang.org/log/1cf98e0db8d30c95362f07d44c47c5363b8c3955)
[2022-08-04T20:04:16-3519aa2-39728f4/netbsd-386-9_0](https://build.golang.org/log/9395fd87814bc4d6344a3bbf72f427637a40ff9d)
[2022-08-04T19:57:25-763f65c-39728f4/netbsd-386-9_0](https://build.golang.org/log/e6a1d9cdf59a37ef2641727a09850df931bf6c76)
[2022-08-04T17:05:18-3e0a503-fb1bfd4/netbsd-amd64-9_0](https://build.golang.org/log/3eab42b91ac882d28586f90d833b6a7adadfaa8e)
[2022-08-04T15:50:11-3e0a503-fcdd099/netbsd-386-9_0](https://build.golang.org/log/f20346aaff38f3501b9586c4bd5bce30beeb544c)
[2022-08-04T15:50:11-3e0a503-44ff9bf/netbsd-amd64-9_0](https://build.golang.org/log/cc32154c3c01fea6178c088a401ce789b8cdfea2)
[2022-08-04T14:58:59-87f47bb-4345620/netbsd-386-9_0](https://build.golang.org/log/1979e42d0a685eaba423a2b72abb5f01467f9d29)
[2022-08-03T21:02:27-8b9a1fb-4345620/netbsd-386-9_0](https://build.golang.org/log/274203fd71753a2475dc09b150bf2223ccafc467)
[2022-08-03T18:07:40-d08f5dc-fcdd099/netbsd-386-9_0](https://build.golang.org/log/0356b836f12656f38f905a8b00195c0dc91f9f9c)
[2022-08-02T18:51:22-74cee27/netbsd-amd64-9_0](https://build.golang.org/log/2a781e06b7500707bc73f861275c17dc176a8b73)
[2022-08-02T18:19:01-d025cce-be59153/netbsd-amd64-9_0](https://build.golang.org/log/d373872d8833fcd7943ed0c44a350b1ed0964171)
[2022-08-02T18:16:22-10cb435-d723df7/netbsd-amd64-9_0](https://build.golang.org/log/52dcc80905a7e4fb69b303d2e10b0853d7a4d249)
[2022-08-02T18:07:14-4d0b383-d723df7/netbsd-386-9_0](https://build.golang.org/log/32e87f0ad4c401d7b6898923db1ccf0e18e96a55)
[2022-08-02T18:07:14-4d0b383-d723df7/netbsd-amd64-9_0](https://build.golang.org/log/c45079cd676d8c78a4fb2e516c7f8c61e99cf932)
[2022-08-02T16:05:48-4d0b383-f2a9f3e/netbsd-amd64-9_0](https://build.golang.org/log/615a1fe11bff97f99eea28f57de7379a57de05ff)
[2022-07-06T19:34:44-fc07039/netbsd-amd64-9_0](https://build.golang.org/log/d63373d5e6980f95caf0effa3552904fde56f521)
[2022-07-06T19:31:41-d69bac6-460a93b/netbsd-386-9_0](https://build.golang.org/log/bf13755fdf4ab98383a4be64a2a0bd647b75bfa3)

</details>

(90 matching logs)

**Comment by bsiegert** [comment]
https://go-review.googlesource.com/c/go/+/315281/ from @tklauser seems highly relevant for this, and it was reverted in https://go-review.googlesource.com/c/go/+/354249 because it resulted in test failures (#48789). Tobias, are there any plans to move forward with that change again?

**Comment by tklauser** [comment]
@bsiegert I haven't looked at it in detail after reverting https://go.dev/cl/315281 and AFAIK I was never able to reproduce the issue on my local NetBSD machine to investigate further.

Given Bryan's comment re. both variants deadlocking I can resend that CL though, so we at least shouldn't hit #13987 on NetBSD.

**Comment by bsiegert** [comment]
Yes, please do!

**Comment by bcmills** [comment]
Huh. https://build.golang.org/log/23bd42ab6595d8c52430290c8bf5d7db3886229d appears to be after the above CL was merged, but it's still showing the hang in a call to `Wait4` instead of Wait6`. 🤔

That call is the one here:
https://cs.opensource.google/go/go/+/master:src/os/exec_unix.go;l=43;drc=a2baae6851a157d662dff7cc508659f66249698a

Is it possible that NetBSD's `wait4` system call is getting confused by a similar PID-reuse race?

**Comment by bcmills** [comment]
Hmm, and that `Wait4` is the same hang reported in #48789. So it seems that using `Wait6` in `blockUntilWaitable` only changes the frequency, not the nature, of the hang.

**Comment by bcmills** [comment]
https://go.dev/cl/436095 and its followup https://go.dev/cl/436656 provide an interesting data-point for `(*os.Process).Wait` on netbsd.

A subprocess was leaked in `TestDoubleStartLeavesPipesOpen`: the process itself exited when its input pipe was closed, but did not result in a `Wait4` to reap the resulting zombie. Somehow, that leaked child process caused all of the `netbsd` builders to fail in a different test (`TestWaitid`, the regression test for #19314) with the following failure mode:
```
--- FAIL: TestWaitid (3.04s)
    exec_posix_test.go:111: os: process already finished
FAIL
```

**Comment by bcmills** [comment]
https://go.dev/cl/439196 provides another interesting data-point. Parallelizing more of the `os/exec` tests dramatically increased the rate of two different failure modes.

―

`netbsd-arm-bsiegert` and `netbsd-arm64-bsiegert` are pretty consistently failing `TestWaitid`, which suggests a kernel bug similar to the macOS bug worked around in #19314.
```
--- FAIL: TestWaitid (3.04s)
    exec_posix_test.go:112: os: process already finished
FAIL
```

―

`netbsd-amd64-9_0` and `netbsd-386-9_0` are pretty consistently hanging in `TestContext` with the `Wait4` hang reported here. (Note that the child process in `TestContext` is terminated with `SIGKILL`, which suggests a possible connection with #42061.)

The goroutine dumps for those failures indicate that `c.Process.Kill` has returned (because the `watchCtx` goroutine is blocked on [the channel send after the call](https://cs.opensource.google/go/go/+/master:src/os/exec/exec.go;l=673;drc=fdea8e21708e9ed04498a7f7b68cbee43005d58d)), although we don't know with what error code, if any, since the stuck `Wait4` call prevents `(*Cmd).Wait` from receiving and returning that error.

```
goroutine 41 [syscall, 3 minutes]:
syscall.Syscall6(0x297e, 0x89526b0, 0x0, 0x8a400f0, 0x0, 0x0, 0x0)
	/tmp/workdir/go/src/syscall/asm_unix_386.s:43 +0x5 fp=0x895265c sp=0x8952658 pc=0x80bd175
syscall.wait4(0x297e, 0x89526b0, 0x0, 0x8a400f0)
	/tmp/workdir/go/src/syscall/zsyscall_netbsd_386.go:34 +0x5b fp=0x8952694 sp=0x895265c pc=0x80bae6b
syscall.Wait4(0x297e, 0x89526d4, 0x0, 0x8a400f0)
	/tmp/workdir/go/src/syscall/syscall_bsd.go:144 +0x3b fp=0x89526b8 sp=0x8952694 pc=0x80b97eb
os.(*Process).wait(0x8b8c090)
	/tmp/workdir/go/src/os/exec_unix.go:43 +0xf3 fp=0x89526f4 sp=0x89526b8 pc=0x80dc5d3
os.(*Process).Wait(...)
	/tmp/workdir/go/src/os/exec.go:132
os/exec.(*Cmd).Wait(0x8b90160)
	/tmp/workdir/go/src/os/exec/exec.go:722 +0x46 fp=0x8952730 sp=0x89526f4 pc=0x81379b6
os/exec_test.TestContext(0x89643c0)
	/tmp/workdir/go/src/os/exec/exec_test.go:922 +0x3e4 fp=0x895279c sp=0x8952730 pc=0x8319e24
testing.tRunner(0x89643c0, 0x83a8c54)
	/tmp/workdir/go/src/testing/testing.go:1501 +0x10d fp=0x89527e4 sp=0x895279c pc=0x812594d
testing.(*T).Run.func1()
	/tmp/workdir/go/src/testing/testing.go:1548 +0x28 fp=0x89527f0 sp=0x89527e4 pc=0x8126738
runtime.goexit()
	/tmp/workdir/go/src/runtime/asm_386.s:1326 +0x1 fp=0x89527f4 sp=0x89527f0 pc=0x80b1141
created by testing.(*T).Run
	/tmp/workdir/go/src/testing/testing.go:1548 +0x379

…

goroutine 148 [chan send, 3 minutes]:
runtime.gopark(0x83a8d54, 0x8b9a270, 0xf, 0x16, 0x2)
	/tmp/workdir/go/src/runtime/proc.go:378 +0xf6 fp=0x8baef24 sp=0x8baef10 pc=0x8082d36
runtime.chansend(0x8b9a240, 0x8baefb0, 0x1, 0x8137919)
	/tmp/workdir/go/src/runtime/chan.go:259 +0x2f6 fp=0x8baef68 sp=0x8baef24 pc=0x804df76
runtime.chansend1(0x8b9a240, 0x8baefb0)
	/tmp/workdir/go/src/runtime/chan.go:145 +0x24 fp=0x8baef80 sp=0x8baef68 pc=0x804dc74
os/exec.(*Cmd).watchCtx(0x8b90160, 0x8b9a240)
	/tmp/workdir/go/src/os/exec/exec.go:673 +0x1a9 fp=0x8baefe4 sp=0x8baef80 pc=0x8137919
os/exec.(*Cmd).Start.func4()
	/tmp/workdir/go/src/os/exec/exec.go:645 +0x28 fp=0x8baeff0 sp=0x8baefe4 pc=0x8137528
runtime.goexit()
	/tmp/workdir/go/src/runtime/asm_386.s:1326 +0x1 fp=0x8baeff4 sp=0x8baeff0 pc=0x80b1141
created by os/exec.(*Cmd).Start
	/tmp/workdir/go/src/os/exec/exec.go:645 +0x885
```

**Comment by bcmills** [comment]
Indeed, merely running `TestContext` in parallel with `TestContextCancel` is enough to reproduce the hang:
```
buildlet-netbsd-amd64-9-0-rncbd5e7a# go test os/exec -timeout=10s -run='TestContext(Cancel)?$' -count=10 -parallel 1
ok      os/exec 0.229s
buildlet-netbsd-amd64-9-0-rncbd5e7a# go test os/exec -timeout=10s -run='TestContext(Cancel)?$' -count=10 -parallel 2
panic: test timed out after 10s

goroutine 20 [running]:
testing.(*M).startAlarm.func1()
        /tmp/workdir/go/src/testing/testing.go:2107 +0x8e
created by time.goFunc
        /tmp/workdir/go/src/time/sleep.go:176 +0x32

goroutine 1 [chan receive]:
testing.tRunner.func1()
        /tmp/workdir/go/src/testing/testing.go:1467 +0x4a5
testing.tRunner(0xc000136680, 0xc00015fb58)
        /tmp/workdir/go/src/testing/testing.go:1507 +0x144
testing.runTests(0xc00011eaa0?, {0x9bbc80, 0x23, 0x23}, {0x0?, 0xc0000061a0?, 0x9c1200?})
        /tmp/workdir/go/src/testing/testing.go:1911 +0x489
testing.(*M).Run(0xc00011eaa0)
        /tmp/workdir/go/src/testing/testing.go:1783 +0x5f7
os/exec_test.TestMain(0xffffffffffffffff?)
        /tmp/workdir/go/src/os/exec/exec_test.go:80 +0xfc
main.main()
        _testmain.go:121 +0x1d3

goroutine 6 [syscall]:
syscall.Syscall6(0x1c1, 0x1a74, 0xc00005ddb4, 0x0, 0xc0001a2000, 0x0, 0x0)
        /tmp/workdir/go/src/syscall/asm_unix_amd64.s:39 +0x5
syscall.wait4(0xc00005ddb8?, 0x40e7c7?, 0x90?, 0x75bfc0?)
        /tmp/workdir/go/src/syscall/zsyscall_netbsd_amd64.go:34 +0x4e
syscall.Wait4(0xc0000221f4?, 0xc00005ddec, 0xc00005ddc8?, 0xc00005de38?)
        /tmp/workdir/go/src/syscall/syscall_bsd.go:144 +0x2b
os.(*Process).wait(0xc0000221e0)
        /tmp/workdir/go/src/os/exec_unix.go:43 +0xbb
os.(*Process).Wait(...)
        /tmp/workdir/go/src/os/exec.go:132
os/exec.(*Cmd).Wait(0xc000144580)
        /tmp/workdir/go/src/os/exec/exec.go:885 +0x45
os/exec_test.TestContext(0xc000136820)
        /tmp/workdir/go/src/os/exec/exec_test.go:946 +0x371
testing.tRunner(0xc000136820, 0x79e5e0)
        /tmp/workdir/go/src/testing/testing.go:1501 +0x10b
created by testing.(*T).Run
        /tmp/workdir/go/src/testing/testing.go:1548 +0x35f

goroutine 8 [chan send]:
os/exec.(*Cmd).watchCtx(0xc000144580, 0xc000100360)
        /tmp/workdir/go/src/os/exec/exec.go:779 +0x49d
created by os/exec.(*Cmd).Start
        /tmp/workdir/go/src/os/exec/exec.go:740 +0xa13
FAIL    os/exec 11.164s
FAIL
```

**Comment by ianlancetaylor** [comment]
There are two different `Kill` functions here (`(*os.Process).Kill` and `syscall.Kill`), so, for clarity: does the failing case call `syscall.Kill` and get an errno result of `ESRCH`?

**Comment by bcmills** [comment]
It looks like it isn't making it to `syscall.Kill` when `TestContext` hangs: the call to `Wait6` returns (prematurely?) and `Wait` calls `(*os.Process).setDone`.

The `ErrProcessDone` comes from here (confirmed with a `println`):
https://cs.opensource.google/go/go/+/master:src/os/exec_unix.go;l=71-73;drc=a2baae6851a157d662dff7cc508659f66249698a

**Comment by ianlancetaylor** [comment]
Thanks.  That makes me think that perhaps `wait6` is not reliable on NetBSD, and that perhaps we should roll back https://go.dev/cl/431855.  It sounds like `wait6` returned, but `wait4` did not immediately succeed.  As far as I can tell that should be impossible.

CC @tklauser

**Comment by bcmills** [comment]
We tried that already, but it only changed the failure rate (see https://github.com/golang/go/issues/50138#issuecomment-1251121891). But perhaps now that we have a more reliable reproducer we can see whether it makes a difference?

I've been trying to reproduce the failure mode in a C program, but so far I haven't had any luck reproducing the early return from `wait6`.

**Comment by bcmills** [comment]
Hrm, but rolling back the `Wait6` change again *does* seem to help with the particular `TestContext` / `TestContextCancel` deadlock. So it's not clear to me whether we're observing the effects of one bug with two variations or two bugs with closely related symptoms. 😩

**Comment by bcmills** [comment]
I left this attempt at a C reproducer running overnight, but by the morning it ended up deadlocked in `fprintf` instead of seeing a spurious return from `wait6`. Apparently there is also some kind of rare race condition in NetBSD's `fprintf` that causes a missed wakeup on a condition variable? 😅

<details>

```C
#include <errno.h>
#include <pthread.h>
#include <sched.h>
#include <signal.h>
#include <stdatomic.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <strings.h>
#include <sys/wait.h>
#include <threads.h>
#include <unistd.h>

void kill_pid(pid_t pid) {
  fprintf(stderr, "killing %d\n", pid);
  if (kill(pid, SIGKILL) != 0) {
    const char *err = strerror(errno);
    fprintf(stderr, "error killing %d: %s\n", pid, err);
    exit(1);
  }
  fprintf(stderr, "killed %d\n", pid);
}

void wait_pid(pid_t pid) {
#ifdef __NetBSD__
  fprintf(stderr, "waiting for %d with wait6\n", pid);
  const pid_t got6 = wait6(P_PID, pid, NULL, WEXITED|WNOWAIT, NULL, NULL);
  if (got6 < 0) {
    const char *err = strerror(errno);
    fprintf(stderr, "error waiting for %d: %s\n", pid, err);
    exit(1);
  } else if (got6 == 0) {
    fprintf(stderr, "wait6 without WNOWAIT for %d returned 0\n", pid);
    exit(1);
  } else if (got6 != pid) {
    fprintf(stderr, "called waitpid(%d, …), but got PID %d\n", pid, got6);
    exit(1);
  }
#else
  fprintf(stderr, "waiting for %d with waitid\n", pid);
  siginfo_t info;
  bzero(&info, sizeof(info));
  if (waitid(P_PID, pid, &info, WEXITED|WNOWAIT) != 0) {
    const char *err = strerror(errno);
    fprintf(stderr, "error waiting for %d: %s\n", pid, err);
    exit(1);
  }
  if (info.si_pid != pid) {
    fprintf(stderr, "waited for %d but got info for %d\n", pid, info.si_pid);
    exit(1);
  }
#endif
  fprintf(stderr, "waited for %d\n", pid);
}

void reap_pid(pid_t pid) {
  fprintf(stderr, "reaping %d\n", pid);
  const pid_t got = waitpid(pid, NULL, 0);
  if (got < 0) {
    const char *err = strerror(errno);
    fprintf(stderr, "error reaping %d: %s\n", pid, err);
    exit(1);
  }
  if (got != pid) {
    fprintf(stderr, "called waitpid(%d, …), but reaped %d\n", pid, got);
  }
  fprintf(stderr, "reaped %d\n", pid);
}


pid_t fork_read_stdin() {
  const pid_t pid = fork();
  if (pid == 0) {
    char buf[8];
    while(read(STDIN_FILENO, &buf, 8) > 0) {}
    exit(0);
  }
  return pid;
}

typedef struct {
  pid_t pid;

  mtx_t mu;
  bool wait_done;
} killThread;

void* do_kill(void* arg) {
  killThread *t = (killThread*)(arg);
  mtx_lock(&t->mu);
  if (t->wait_done) {
    fprintf(stderr, "wait for %d completed before it was killed\n", t->pid);
    exit(1);
  } else {
    kill_pid(t->pid);
  }
  mtx_unlock(&t->mu);
  return NULL;
}

void* do_kill_wait_reap(void* unused) {
  killThread t = {.wait_done = false};
  mtx_init(&t.mu, mtx_plain);

  t.pid = fork_read_stdin();

  pthread_t kill_thread;
  if (pthread_create(&kill_thread, NULL, do_kill, &t) != 0) {
    perror("starting thread3: ");
    exit(1);
  }

  wait_pid(t.pid);
  mtx_lock(&t.mu);
  t.wait_done = true;
  mtx_unlock(&t.mu);

  reap_pid(t.pid);
  pthread_join(kill_thread, NULL);
  return NULL;
}


int main(int argc, char *argv) {
  pthread_t thread1;
  if (pthread_create(&thread1, NULL, do_kill_wait_reap, NULL) != 0) {
    perror("starting thread1: ");
    exit(1);
  }

  pthread_t thread2;
  if (pthread_create(&thread2, NULL, do_kill_wait_reap, NULL) != 0) {
    perror("starting thread2: ");
    exit(1);
  }

  pthread_join(thread2, NULL);
  pthread_join(thread1, NULL);
  return 0;
}
```

```
(gdb) i thr
  Id   Target Id                Frame
* 1    LWP 5 of process 2552 "" 0x000079eef42a1f7a in ___lwp_park60 () from /usr/lib/libc.so.12
  2    LWP 3 of process 2552 "" 0x000079eef424277a in wait6 () from /usr/lib/libc.so.12
  3    LWP 1 of process 2552 "" 0x000079eef42a1f7a in ___lwp_park60 () from /usr/lib/libc.so.12
(gdb) thr 1
[Switching to thread 1 (LWP 5 of process 2552)]
#0  0x000079eef42a1f7a in ___lwp_park60 () from /usr/lib/libc.so.12
(gdb) bt
#0  0x000079eef42a1f7a in ___lwp_park60 () from /usr/lib/libc.so.12
#1  0x000079eef480a71f in pthread_cond_timedwait () from /usr/lib/libpthread.so.1
#2  0x000079eef4367e05 in __flockfile_internal () from /usr/lib/libc.so.12
#3  0x000079eef4349cb6 in vfprintf () from /usr/lib/libc.so.12
#4  0x000079eef4344114 in fprintf () from /usr/lib/libc.so.12
#5  0x0000000000400dfe in kill_pid ()
#6  0x00000000004010cf in do_kill ()
#7  0x000079eef480c072 in ?? () from /usr/lib/libpthread.so.1
#8  0x000079eef4287af0 in ?? () from /usr/lib/libc.so.12
#9  0x0000000000000000 in ?? ()
(gdb) thr 2
[Switching to thread 2 (LWP 3 of process 2552)]
#0  0x000079eef424277a in wait6 () from /usr/lib/libc.so.12
(gdb) bt
#0  0x000079eef424277a in wait6 () from /usr/lib/libc.so.12
#1  0x0000000000400eb2 in wait_pid ()
#2  0x0000000000401184 in do_kill_wait_reap ()
#3  0x000079eef480c072 in ?? () from /usr/lib/libpthread.so.1
#4  0x000079eef4287af0 in ?? () from /usr/lib/libc.so.12
#5  0x0000000000000000 in ?? ()
(gdb) thr 3
[Switching to thread 3 (LWP 1 of process 2552)]
#0  0x000079eef42a1f7a in ___lwp_park60 () from /usr/lib/libc.so.12
(gdb) bt
#0  0x000079eef42a1f7a in ___lwp_park60 () from /usr/lib/libc.so.12
#1  0x000079eef480a71f in pthread_cond_timedwait () from /usr/lib/libpthread.so.1
#2  0x000079eef480c2c8 in pthread_join () from /usr/lib/libpthread.so.1
#3  0x0000000000401250 in main ()
```

</details>

**Comment by bcmills** [comment]
Attaching and detaching `gdb` from the aforementioned deadlocked process un-deadlocked it. Smells like another kernel bug, but it's not clear to me whether it's related to the bug affecting `wait6`. 🤦‍♂️

**Comment by bcmills** [comment]
`kdump` output for the following test run is at https://gist.github.com/bcmills/fc08fec14bdce9f3a98692f7cf73efde.

<details>

```
buildlet-netbsd-amd64-9-0-rnc24fbc7# ktrace -i ./exec.test -test.run='TestContext(Cancel)?$' -test.timeout=10s -test.v
=== RUN   TestContext
=== PAUSE TestContext
=== RUN   TestContextCancel
=== PAUSE TestContextCancel
=== CONT  TestContext
=== CONT  TestContextCancel
killed 23884
wait6 returned for 25497
wait6 returned for 23884
    exec_test.go:1011: exit status: signal: killed
--- PASS: TestContextCancel (0.00s)
panic: test timed out after 10s

goroutine 10 [running]:
testing.(*M).startAlarm.func1()
        /tmp/workdir/go/src/testing/testing.go:2107 +0x8e
created by time.goFunc
        /tmp/workdir/go/src/time/sleep.go:176 +0x32

goroutine 1 [chan receive]:
testing.tRunner.func1()
        /tmp/workdir/go/src/testing/testing.go:1467 +0x4a5
testing.tRunner(0xc000132680, 0xc00015bb58)
        /tmp/workdir/go/src/testing/testing.go:1507 +0x144
testing.runTests(0xc00011ea00?, {0x9b8c80, 0x23, 0x23}, {0x0?, 0xc0000061a0?, 0x9be200?})
        /tmp/workdir/go/src/testing/testing.go:1911 +0x489
testing.(*M).Run(0xc00011ea00)
        /tmp/workdir/go/src/testing/testing.go:1783 +0x5f7
os/exec_test.TestMain(0xffffffffffffffff?)
        /tmp/workdir/go/src/os/exec/exec_test.go:81 +0xfc
main.main()
        _testmain.go:121 +0x1d3

goroutine 6 [syscall]:
syscall.Syscall6(0x1c1, 0x6399, 0xc00005bdb4, 0x0, 0xc000168000, 0x0, 0x0)
        /tmp/workdir/go/src/syscall/asm_unix_amd64.s:39 +0x5
syscall.wait4(0xc00005bdb8?, 0x40e7c7?, 0x90?, 0x758fc0?)
        /tmp/workdir/go/src/syscall/zsyscall_netbsd_amd64.go:34 +0x4e
syscall.Wait4(0xc000024194?, 0xc00005bdec, 0xc00005bdc8?, 0xc00005be38?)
        /tmp/workdir/go/src/syscall/syscall_bsd.go:144 +0x2b
os.(*Process).wait(0xc000024180)
        /tmp/workdir/go/src/os/exec_unix.go:43 +0xbb
os.(*Process).Wait(...)
        /tmp/workdir/go/src/os/exec.go:132
os/exec.(*Cmd).Wait(0xc000140420)
        /tmp/workdir/go/src/os/exec/exec.go:882 +0x45
os/exec_test.TestContext(0xc000132820)
        /tmp/workdir/go/src/os/exec/exec_test.go:947 +0x371
testing.tRunner(0xc000132820, 0x79b5d8)
        /tmp/workdir/go/src/testing/testing.go:1501 +0x10b
created by testing.(*T).Run
        /tmp/workdir/go/src/testing/testing.go:1548 +0x35f

goroutine 8 [chan send]:
os/exec.(*Cmd).watchCtx(0xc000140420, 0xc000100360)
        /tmp/workdir/go/src/os/exec/exec.go:779 +0x405
created by os/exec.(*Cmd).Start
        /tmp/workdir/go/src/os/exec/exec.go:740 +0xa13
```

</details>

**Comment by bcmills** [comment]
https://go.dev/cl/442478 seems like a plausible fix for this issue. Waiting to see if there's more.

The TryBots for that change hit a deadlock that looks plausibly related to the missed `fprintf` wakeup from the C program I tried yesterday (https://github.com/golang/go/issues/50138#issuecomment-1276074607). If that's all that is left I suggest that we report it as a separate issue, since it has a somewhat different symptom.

**Comment by bcmills** [comment]
I've filed the new failure mode separately as #56180.

**Comment by riastradh** [comment]
> I left this attempt at a C reproducer running overnight on a `netbsd-amd64-9_0` gomote instance, but by the morning it ended up deadlocked in `fprintf` instead of seeing a spurious return from `wait6`. Apparently there is also some kind of rare race condition in NetBSD's `fprintf` that causes a missed wakeup on a condition variable? 😅

The fprintf issue you hit may be a bug in the interaction between libpthread and the dynamic loader rtld which we have since fixed in HEAD and netbsd-9 (but not in netbsd-9.0):

https://nxr.netbsd.org/xref/src/lib/libpthread/pthread.c?r=1.181#418

Please let us know if you can still reproduce it on a current system.

> https://go.dev/cl/442478 seems like a plausible fix for this issue. Waiting to see if there's more.

Surely that wouldn't affect the deadlock you saw with wait4, would it?  If Go uses wait4 instead of wait6, do you still see the deadlock?  (Forgive me if I missed something -- there have been a lot of updates in quick succession which I didn't follow all of.)

**Comment by bcmills** [comment]
The `wait4` version of the deadlock is plausibly #13987, depending how quickly NetBSD recycles PIDs.

**Comment by bcmills** [comment]
Updating to a more recent NetBSD release is #54773.

**Comment by riastradh** [comment]
> Updating to a more recent NetBSD release is #54773.

Unfortunately the libpthread/rtld fix didn't make it into 9.3!  I didn't realize it until a couple days after 9.3 went out, sorry.  But if you can reproduce the C program's fprintf hang, just a regular NetBSD install on a VM without all the golang test harness, that would be helpful -- especially if you can do it on a system with the debug.tgz (or debug.tar.xz) set installed so we get full stack traces.  I'll see if I can reproduce it, but half an hour of running it, both on a NetBSD<=9.3 library without the fix and on a current library with the fix, hasn't turned anything up yet.

**Comment by bsiegert** [comment]
Instead of updating the builder VMs to 9.3, should we be updating them to a more recent snapshot of NetBSD-9 then?

**Comment by bcmills** [comment]
The new failure mode in #56180 is consistent with the aforementioned `libpthread` bug, and the remaining failures on the builders after CL 442478 all seem to plausibly match that failure mode.

I'm going to close this issue as “one bug fixed”, with the remaining hang tracked in #56180. (If there are other failure modes after the builders are upgraded in #54773, we can open new issues for those.)
